// react-nativescript-navigation/dist/native-stack/views/FrameNavigatorView.js
/**
 * A reimplementation of React Native Screens for React NativeScript.
 * @see https://github.com/software-mansion/react-native-screens/blob/73959abc975b5718e683d39f1452ec0bb4d5f475/native-stack/views/NativeStackView.tsx
 */
import * as React from 'react';
// import { View, StyleSheet, Platform } from 'react-native';
import { ScreenStack, Screen as ScreenComponent, } from '../../react-nativescript-screens/screens';
import { StackActions, } from '@react-navigation/core';
import HeaderConfig from './HeaderConfig';
import { isAndroid } from "@nativescript/core";
const Screen = ScreenComponent;
/**
 *
 * @see https://github.com/software-mansion/react-native-screens/blob/73959abc975b5718e683d39f1452ec0bb4d5f475/native-stack/views/Frame.tsx#L27
 */
export default function FrameNavigatorView(_a) {
    // const { colors } = useTheme();
    // console.log(`[Frame] ${JSON.stringify(state.routes.map(route => route.key))}`);
    var { state, navigation, descriptors, style } = _a, rest = __rest(_a, ["state", "navigation", "descriptors", "style"]);
    return (
    // Frame
    React.createElement(ScreenStack, Object.assign({}, rest, { style: Object.assign(Object.assign({}, styles.container), style) }), state.routes.map((route, index, self) => {
        const { options, render: renderScene } = descriptors[route.key];
        const { gestureEnabled = true, stackPresentation = 'push', stackAnimation, contentStyle, headerShown, } = options;
        const active = index === self.length - 1;
        return (
        // Page
        React.createElement(Screen, { key: route.key, active: active ? 1 : 0, style: styles.fill, actionBarHidden: !headerShown, gestureEnabled: isAndroid ? false : gestureEnabled, stackPresentation: stackPresentation, stackAnimation: stackAnimation, onWillDisappear: (args) => {
                // console.log(`[Screen.${route.key} ${args.object}] 'willDisappear'.`);
            }, onDidDisappear: (args) => {
                var _a;
                const page = args.object;
                const frame = page.frame;
                if ((_a = frame === null || frame === void 0 ? void 0 : frame._currentEntry) === null || _a === void 0 ? void 0 : _a.resolvedPage) {
                    // console.log(`[Screen.${route.key} ${args.object}] 'didDisappear'. Still currentPage.`);
                    // "First" screen fires this upon disappearing when navigating forward to "Second" via React Navigation.
                    // "Second" screen fires this upon disappearing when navigating backward to "First" via React Navigation.
                    // "Second" screen fires this upon disappearing when navigating backward to "First" via user.
                }
                else {
                    // console.log(`[Screen.${route.key} ${args.object}] 'didDisappear'. No longer currentPage.`);
                }
                navigation.emit({
                    type: 'didDisappear',
                    target: route.key,
                });
                /**
                 * When navigation.navigate('second') is called, a route is pushed to state.routes.
                 * This causes RNS to call parent.insert(child), which involves frame.navigate({ create(): { return page; } });
                 *
                 * When navigation.goBack() is called, a route is popped from state.routes.
                 * This causes RNS to call parent.remove(child), which involves frame.goBack();
                 *
                 * So, in those two cases, there's no need to dispatch a "POP" action to bring state.routes back in sync
                 * with the stack maintained by the Frame. Because the ReactChildren are driving the navigation state.
                 *
                 * The worrying cases come from Frame's side:
                 * – Android back button (pop)
                 * – Right swipe (pop)
                 * – Press "back" navigation button (pop)
                 *
                 * If this were purely RNS (no React Navigation), we'd need to capture (and more importantly, discern) these
                 * events and manually pop the necessary Page(s) off our Frame's array of ReactChildren.
                 *
                 * The same goes for React Navigation, with the subtlety that said array (state.routes) is managed internally
                 * by React Navigation. To pop from it, we must dispatch a "POP" action.
                 *
                 * However, we will also need to rely on some implicit behaviour in the RNS registry. If ever
                 * parent.remove(child) is called for a child that is not the currentPage of the frame, we must be careful
                 * _not_ to call frame.goBack() (because it has already been done); instead, we must simply no-op.
                 *
                 *
                 *
                 */
            }, onWillAppear: (args) => {
                // console.log(`[Screen.${route.key} ${args.object}] 'willAppear'.`);
            }, onDidAppear: (args) => {
                var _a;
                // Software Mansion's Frame chooses to emit before the dispatch.
                navigation.emit({
                    type: 'didAppear',
                    target: route.key,
                });
                const page = args.object;
                const frame = page.frame;
                if (((_a = frame === null || frame === void 0 ? void 0 : frame._currentEntry) === null || _a === void 0 ? void 0 : _a.resolvedPage) === page) {
                    // "Second" screen fires this upon appearing when being navigating forward to from "First" via React Navigation.
                    // "First" screen fires this upon appearing when being navigating backward to from "Second" via React Navigation.
                    // "First" screen fires this upon appearing when being navigating backward to from "Second" via user.
                    // console.log(`[Screen.${route.key} ${args.object}] 'didAppear', becoming currentPage. active: ${active}`);
                    if (!active) {
                        const topRoute = state.routes[state.routes.length - 1];
                        navigation.dispatch(Object.assign(Object.assign({}, StackActions.pop()), { source: topRoute.key, target: state.key }));
                    }
                }
                else {
                    // console.log(`[Screen.${route.key} ${args.object}] 'didAppear', but not becoming currentPage.`);
                }
            } },
            React.createElement(HeaderConfig, Object.assign({}, options, { route: route })),
            React.createElement("flexboxLayout", { nodeRole: "content", style: Object.assign(Object.assign({}, styles.flexContainer), contentStyle) }, renderScene())));
    })));
}
const styles = {
    fill: {
        width: "100%",
        height: "100%",
    },
    container: {
        width: "100%",
        height: "100%",
    },
    flexContainer: {
        flexGrow: 1,
        flexDirection: "column",
        width: "100%",
        height: "100%",
    },
};
